harfbuzz/src/buffer.rs

```diff
@@ -0,0 +1,308 @@
+///   behaviour. Languages are orthogonal to the scripts, and though they
+///   are related, they are different concepts and should not be confused
+///   with each other. This is controlled via the [`set_language`] method
+///   on `Buffer`.
+///
+/// Additionally, Harfbuzz can attempt to infer the values for these
+/// properties using the [`guess_segment_properties`] method on `Buffer`:
+///
+/// ```
+/// # use harfbuzz::{Buffer, Direction, sys};
+/// let mut b = Buffer::with("مساء الخير");
+/// b.guess_segment_properties();
+/// assert_eq!(b.get_direction(), Direction::RTL);
+/// assert_eq!(b.get_script(), sys::HB_SCRIPT_ARABIC);
+/// ```
+///
+/// [`set_direction`]: #method.set_direction
+/// [`set_script`]: #method.set_script
+/// [`set_language`]: #method.set_language
+/// [`guess_segment_properties`]: #method.guess_segment_properties
+pub struct Buffer {
+    /// The underlying `hb_buffer_t` from the `harfbuzz-sys` crate.
+    ///
+    /// This isn't commonly needed unless interfacing directly with
+    /// functions from the `harfbuzz-sys` crate that haven't been
+    /// safely exposed.
+    pub raw: *mut sys::hb_buffer_t,
+}
+
+impl Buffer {
+    /// Create a new, empty buffer.
+    ///
+    /// ```
+    /// # use harfbuzz::Buffer;
+    /// let b = Buffer::new();
+    /// assert!(b.is_empty());
+    /// ```
+    pub fn new() -> Self {
+        Buffer::default()
+    }
+
+    /// Create a new buffer with the given text.
+    pub fn with(text: &str) -> Self {
+        let mut b = Buffer::new();
+        b.add_str(text);
+        b
+    }
+
+    /// Create a new, empty buffer with the specified capacity.
+    pub fn with_capacity(capacity: usize) -> Self {
+        let mut b = Buffer::default();
+        b.reserve(capacity);
+        b
+    }
+
+    /// Add UTF-8 encoded text to the buffer.
+    pub fn add_str(&mut self, text: &str) {
+        unsafe {
+            sys::hb_buffer_add_utf8(
+                self.raw,
+                text.as_ptr() as *const std::os::raw::c_char,
+                text.len() as std::os::raw::c_int,
+                0,
+                text.len() as std::os::raw::c_int,
+            )
+        };
+    }
+
+    /// Append part of the contents of another buffer to this one.
+    ///
+    /// ```
+    /// # use harfbuzz::Buffer;
+    /// let mut b1 = Buffer::with("butter");
+    /// let b2 = Buffer::with("fly");
+    /// b1.append(&b2, 0, 3);
+    /// assert_eq!(b1.len(), "butterfly".len());
+    /// ```
+    pub fn append(&mut self, other: &Buffer, start: usize, end: usize) {
+        unsafe {
+            sys::hb_buffer_append(
+                self.raw,
+                other.raw,
+                start as std::os::raw::c_uint,
+                end as std::os::raw::c_uint,
+            )
+        };
+    }
+
+    /// Throw away text stored in the buffer, but maintain the
+    /// currently configured options like direction and script.
```

Comment:
`hb_buffer_clear_contents` actually clears the segment properties as well.

Reply:
Looking at the harfbuzz code, I see that the difference is that `reset` resets the unicode functions (which I haven't exposed yet), the buffer flags (which I don't yet understand well), and the replacement character, which isn't needed from Rust in the current API as a `&str` will be valid UTF-8 and shouldn't need the replacement character.

This probably also means that I should submit a patch upstream to improve the harfbuzz docs on this point.